### ISSUE TITLE
LTO breaks TRIM_UNUSED_KSYMS when symbols are only referenced from inline assembly

### ISSUE THREAD

**Issue by samitolvanen** [ClangBuiltLinux/linux#1369]
@nickdesaulniers  reports that x86_64 defconfig + `CONFIG_LTO_CLANG_THIN` + `CONFIG_TRIM_UNUSED_KSYMS` currently fails to build with the following modpost error:
```
  MODPOST modules-only.symvers
ERROR: modpost: "ex_handler_rdmsr_unsafe" [drivers/thermal/intel/x86_pkg_temp_thermal.ko] undefined!
ERROR: modpost: "ex_handler_wrmsr_unsafe" [drivers/thermal/intel/x86_pkg_temp_thermal.ko] undefined!
make[2]: *** [../scripts/Makefile.modpost:150: modules-only.symvers] Error 1
```

This happens because the list of undefined symbols in `x86_pkg_temp_thermal.ko` is actually determined from the LLVM bitcode in `x86_pg_temp_thermal.o`. While the bitcode does contain references to these symbols, they are in inline assembly, which `llvm-nm` ignores.

An obvious solution would be to generate the list of undefined symbols only after we have compiled the bitcode into `.lto.o`, but that's complicated by the fact that the `undefined_syms` command is run before the module is linked into `.lto.o`. Here's the relevant part from `scripts/Makefile.build`:
```Make
ifdef CONFIG_TRIM_UNUSED_KSYMS
...
# List module undefined symbols
undefined_syms = $(NM) $< | $(AWK) '$$1 == "U" { printf("%s%s", x++ ? " " : "", $$2) }';
endif
...
cmd_mod = { \
        echo $(if $($*-objs)$($*-y)$($*-m), $(addprefix $(obj)/, $($*-objs) $($*-y) $($*-m)), $(@:.mod=.o)); \
        $(undefined_syms) echo; \
        } > $@

$(obj)/%.mod: $(obj)/%.o FORCE
        $(call if_changed,mod)
```

In order to fix this, I think we would have to either postpone `undefined_syms` until after `.lto.o` has been compiled, or compile `.lto.o` earlier.

**Comment by nathanchance** [comment]
Is this series relevant? https://lore.kernel.org/r/20210309151737.345722-1-masahiroy@kernel.org/

**Comment by samitolvanen** [comment]
That series doesn't seem to change how the list of undefined symbols is generated. We would still have to look at `.lto.o` to make sure none of the symbols are missing.

**Comment by samitolvanen** [comment]
This quick patch fixes the issue for me:
```diff
diff --git a/scripts/Makefile.build b/scripts/Makefile.build
index 5e39b0517186..784546d672f5 100644
--- a/scripts/Makefile.build
+++ b/scripts/Makefile.build
@@ -272,12 +272,28 @@ $(obj)/%.o: $(src)/%.c $(recordmcount_source) $(objtool_dep) FORCE
        $(call if_changed_rule,cc_o_c)
        $(call cmd,force_checksrc)
 
+ifdef CONFIG_LTO_CLANG
+# With CONFIG_LTO_CLANG, .o files might be LLVM bitcode, so we need to run
+# LTO to compile them into native code before further processing
+prelink-ext := .lto
+
+quiet_cmd_cc_lto_link_modules = LTO [M] $@
+cmd_cc_lto_link_modules =                                              \
+       $(LD) $(ld_flags) -r -o $@                                      \
+               $(shell [ -s $(@:.lto.o=.o.symversions) ] &&            \
+                       echo -T $(@:.lto.o=.o.symversions))             \
+               --whole-archive $^
+
+$(obj)/%.lto.o: $(obj)/%.o
+       $(call if_changed,cc_lto_link_modules)
+endif
+
 cmd_mod = { \
        echo $(if $($*-objs)$($*-y)$($*-m), $(addprefix $(obj)/, $($*-objs) $($*-y) $($*-m)), $(@:.mod=.o)); \
        $(undefined_syms) echo; \
        } > $@
 
-$(obj)/%.mod: $(obj)/%.o FORCE
+$(obj)/%.mod: $(obj)/%$(prelink-ext).o FORCE
        $(call if_changed,mod)
 
 quiet_cmd_cc_lst_c = MKLST   $@
diff --git a/scripts/Makefile.modpost b/scripts/Makefile.modpost
index c383ba33d837..06ee7678ba83 100644
--- a/scripts/Makefile.modpost
+++ b/scripts/Makefile.modpost
@@ -41,9 +41,6 @@ __modpost:
 include include/config/auto.conf
 include $(srctree)/scripts/Kbuild.include
 
-# for ld_flags
-include $(srctree)/scripts/Makefile.lib
-
 MODPOST = scripts/mod/modpost                                                          \
        $(if $(CONFIG_MODVERSIONS),-m)                                                  \
        $(if $(CONFIG_MODULE_SRCVERSION_ALL),-a)                                        \
@@ -119,19 +116,8 @@ $(input-symdump):
        @echo >&2 '         You may get many unresolved symbol warnings.'
 
 ifdef CONFIG_LTO_CLANG
-# With CONFIG_LTO_CLANG, .o files might be LLVM bitcode, so we need to run
-# LTO to compile them into native code before running modpost
+# With CONFIG_LTO_CLANG, .o files might be LLVM bitcode, so use .lto.o for modpost
 prelink-ext := .lto
-
-quiet_cmd_cc_lto_link_modules = LTO [M] $@
-cmd_cc_lto_link_modules =                                              \
-       $(LD) $(ld_flags) -r -o $@                                      \
-               $(shell [ -s $(@:.lto.o=.o.symversions) ] &&            \
-                       echo -T $(@:.lto.o=.o.symversions))             \
-               --whole-archive $^
-
-%.lto.o: %.o
-       $(call if_changed,cc_lto_link_modules)
 endif
 
 modules := $(sort $(shell cat $(MODORDER)))
```
Both defconfig and allmodconfig seem to build with ThinLTO and `TRIM_UNUSED_KSYMS` after I apply the patch, but this is going to need a lot more testing.

**Comment by nickdesaulniers** [comment]
@MaskRay mentioned in today's meeting that llvm-nm parses module level inline asm, but not function level inline asm (IIUC). If so, that seems curious.

**Comment by MaskRay** [comment]
The IR symbol table tracks symbols referenced from module level inline asm, but not symbols referenced by function scope inline asm. The issue may be that the LTO library doesn't know that some symbols are referenced by regular (ELF) object files and should be retained, so such symbols are incorrectly discarded.

I thought about whether we should track inline asm, but this requires another iteration over all instructions in a module, so I feel that this is unnecessary.

See https://www.openwall.com/lists/musl/2020/12/27/25 for another instance. If this comes up again, then this issue is probably annoying enough that we should address it.

**Comment by ms178** [comment]
> This quick patch fixes the issue for me:
> 
> ```diff
> diff --git a/scripts/Makefile.build b/scripts/Makefile.build
> index 5e39b0517186..784546d672f5 100644
> --- a/scripts/Makefile.build
> +++ b/scripts/Makefile.build
> @@ -272,12 +272,28 @@ $(obj)/%.o: $(src)/%.c $(recordmcount_source) $(objtool_dep) FORCE
>         $(call if_changed_rule,cc_o_c)
>         $(call cmd,force_checksrc)
>  
> +ifdef CONFIG_LTO_CLANG
> +# With CONFIG_LTO_CLANG, .o files might be LLVM bitcode, so we need to run
> +# LTO to compile them into native code before further processing
> +prelink-ext := .lto
> +
> +quiet_cmd_cc_lto_link_modules = LTO [M] $@
> +cmd_cc_lto_link_modules =                                              \
> +       $(LD) $(ld_flags) -r -o $@                                      \
> +               $(shell [ -s $(@:.lto.o=.o.symversions) ] &&            \
> +                       echo -T $(@:.lto.o=.o.symversions))             \
> +               --whole-archive $^
> +
> +$(obj)/%.lto.o: $(obj)/%.o
> +       $(call if_changed,cc_lto_link_modules)
> +endif
> +
>  cmd_mod = { \
>         echo $(if $($*-objs)$($*-y)$($*-m), $(addprefix $(obj)/, $($*-objs) $($*-y) $($*-m)), $(@:.mod=.o)); \
>         $(undefined_syms) echo; \
>         } > $@
>  
> -$(obj)/%.mod: $(obj)/%.o FORCE
> +$(obj)/%.mod: $(obj)/%$(prelink-ext).o FORCE
>         $(call if_changed,mod)
>  
>  quiet_cmd_cc_lst_c = MKLST   $@
> diff --git a/scripts/Makefile.modpost b/scripts/Makefile.modpost
> index c383ba33d837..06ee7678ba83 100644
> --- a/scripts/Makefile.modpost
> +++ b/scripts/Makefile.modpost
> @@ -41,9 +41,6 @@ __modpost:
>  include include/config/auto.conf
>  include $(srctree)/scripts/Kbuild.include
>  
> -# for ld_flags
> -include $(srctree)/scripts/Makefile.lib
> -
>  MODPOST = scripts/mod/modpost                                                          \
>         $(if $(CONFIG_MODVERSIONS),-m)                                                  \
>         $(if $(CONFIG_MODULE_SRCVERSION_ALL),-a)                                        \
> @@ -119,19 +116,8 @@ $(input-symdump):
>         @echo >&2 '         You may get many unresolved symbol warnings.'
>  
>  ifdef CONFIG_LTO_CLANG
> -# With CONFIG_LTO_CLANG, .o files might be LLVM bitcode, so we need to run
> -# LTO to compile them into native code before running modpost
> +# With CONFIG_LTO_CLANG, .o files might be LLVM bitcode, so use .lto.o for modpost
>  prelink-ext := .lto
> -
> -quiet_cmd_cc_lto_link_modules = LTO [M] $@
> -cmd_cc_lto_link_modules =                                              \
> -       $(LD) $(ld_flags) -r -o $@                                      \
> -               $(shell [ -s $(@:.lto.o=.o.symversions) ] &&            \
> -                       echo -T $(@:.lto.o=.o.symversions))             \
> -               --whole-archive $^
> -
> -%.lto.o: %.o
> -       $(call if_changed,cc_lto_link_modules)
>  endif
>  
>  modules := $(sort $(shell cat $(MODORDER)))
> ```
> 
> Both defconfig and allmodconfig seem to build with ThinLTO and `TRIM_UNUSED_KSYMS` after I apply the patch, but this is going to need a lot more testing.

Thanks, I've applied your patch manually to my 5.12.3-Xanmod-Kernel (which I use with the Speculative Page Fault series and Cluster-Topology Scheduler patches on top) and it compiles and boots fine now. While testing this patch, I've noticed some problems with copy and pasting text between the browser and the console or Dolphin file manager, I get the following new dmesg-entries:

[  +0,000096] alloc_contig_range: [100bc0, 100be0) PFNs busy
[  +0,000082] alloc_contig_range: [100bc0, 100be0) PFNs busy
[  +0,000071] alloc_contig_range: [100bc0, 100be0) PFNs busy
[  +0,000071] alloc_contig_range: [100bc0, 100be0) PFNs busy
[  +0,000069] alloc_contig_range: [100bc0, 100be0) PFNs busy
[  +0,000064] alloc_contig_range: [100bc0, 100be0) PFNs busy
[  +0,000064] alloc_contig_range: [100bc0, 100be0) PFNs busy
[  +6,727699] alloc_contig_range: [100bd8, 100be0) PFNs busy
[ +19,324885] alloc_contig_range: [100bda, 100bdc) PFNs busy
[  +0,345130] alloc_contig_range: [100bda, 100bdc) PFNs busy
[  +0,298596] alloc_contig_range: [100bda, 100bdc) PFNs busy

Here are the config and Makefile used, I use the LLVM/Clang-12 toolchain from Oibaf's PPA:
[config.txt](https://github.com/ClangBuiltLinux/linux/files/6465579/config.txt)
[Makefile.txt](https://github.com/ClangBuiltLinux/linux/files/6465580/Makefile.txt)

Notice, that I also set some compiler flags manually in the Makefile (using polly).

Edit: After a couple of reboots I haven't seen this issue again.

**Comment by ms178** [comment]
By the way, the performance difference between a GCC vs. Clang-compiled Kernel (both equally optimized) is around 5% in favor of the Clang-built Kernel in the WebGL Aquarium benchmark with 10.000 fishes (which I use for a quick performance test): 98 vs 103 fps (https://webglsamples.org/aquarium/aquarium.html)

**Comment by samitolvanen** [comment]
> Thanks, I've applied your patch manually to my 5.12.3-Xanmod-Kernel (which I use with the Speculative Page Fault series and Cluster-Topology Scheduler patches on top) and it compiles and boots fine now. 

Great, thank you for testing!

**Comment by ms178** [comment]
@samitolvanen 

I've just tried Full LTO with TRIM_UNUSED_KSYSMS and your patch, much to my surprise it compiled fine (albeit the compressed deb-file is roughly 10% bigger than the ThinLTO variant). Unfortunately the Kernel would not boot and hangs between GRUB and initializing the kernel (I've also hit this symptom with GCC LTO by the way).

**Comment by samitolvanen** [comment]
https://lore.kernel.org/lkml/20210716204545.3536354-1-samitolvanen@google.com/

**Comment by solbjorn** [comment]
For the version that was pushed 11 hours ago:

```
Reviewed-by: Alexander Lobakin <alobakin@pm.me>
Tested-by: Alexander Lobakin <alobakin@pm.me>
```

**Comment by samitolvanen** [comment]
> For the version that was pushed 11 hours ago:
> 
> ```
> Reviewed-by: Alexander Lobakin <alobakin@pm.me>
> Tested-by: Alexander Lobakin <alobakin@pm.me>
> ```

I fixed one more typo in a comment and will include these in v2. Thanks for testing!

**Comment by samitolvanen** [comment]
Accepted: https://git.kernel.org/pub/scm/linux/kernel/git/masahiroy/linux-kbuild.git/commit/?h=for-next&id=2b1d7fc054677a29abad15b3b32dad93df6eb9cf

**Comment by nickdesaulniers** [comment]
850ded46c64299f04d15e39caaba21963fb966f8
